File: Misc/127_lc_1615.java
/*
 * 1615. Maximal Network Rank
 */

 import java.util.*;

 class Solution {
     public int maximalNetworkRank(int n, int[][] roads) {
         Set<String> existingRoads = new HashSet<>();
         int[] nodeVsNeighbourCount = new int[n];
 
         for (int[] road : roads) {
             existingRoads.add(road[0] + "," + road[1]);
             existingRoads.add(road[1] + "," + road[0]); // Add both directions
             nodeVsNeighbourCount[road[0]]++;
             nodeVsNeighbourCount[road[1]]++;
         }
 
         int res = 0;
         for (int city1 = 0; city1 < n - 1; city1++) {
             for (int city2 = city1 + 1; city2 < n; city2++) {
                 int count = nodeVsNeighbourCount[city1] + nodeVsNeighbourCount[city2];
                 if (existingRoads.contains(city1 + "," + city2)) {
                     count--; // Subtract 1 if there is a direct road
                 }
                 res = Math.max(res, count);
             }
         }
 
         return res;
     }
 }
 

File: Misc/127_lc_1615.py
"""
1615. Maximal Network Rank
"""

from typing import List


class Solution:
    def maximalNetworkRank(self, n: int, roads: List[List[int]]) -> int:
        existingRoads = set()
        nodeVsNeighbourCount = [0 for _ in range(n)]
        for road in roads:
            existingRoads.add((road[0], road[1]))
            nodeVsNeighbourCount[road[0]] += 1
            nodeVsNeighbourCount[road[1]] += 1

        res = 0
        for city1 in range(n - 1):
            for city2 in range(city1 + 1, n):
                count = nodeVsNeighbourCount[city1] + nodeVsNeighbourCount[city2]
                if (city1, city2) in existingRoads or (city2, city1) in existingRoads:
                    count -= 1

                res = max(res, count)

        return res
